elephant/asset/asset.py

```diff
@@ -410,224 +429,247 @@ def _interpolate_signals(signals, sampling_times, verbose=False):
-    sequence_sorted = list(range(d, 0, -1))
-    input_order = tuple(sequence_sorted)  # fixed
-    while sequence_sorted[0] != n + 1:
-        for last_element in range(1, sequence_sorted[-2] + 1):
-            sequence_sorted[-1] = last_element
-            yield tuple(sequence_sorted)
-        increment_id = d - 2
-        while increment_id > 0 and sequence_sorted[increment_id - 1] == \
-                sequence_sorted[increment_id]:
-            increment_id -= 1
-        sequence_sorted[increment_id + 1:] = input_order[increment_id + 1:]
-        sequence_sorted[increment_id] += 1
-
-
-def _jsf_uniform_orderstat_3d(u, n, verbose=False):
-    r"""
-    Considered n independent random variables X1, X2, ..., Xn all having
-    uniform distribution in the interval (0, 1):
+class _JSFUniformOrderStat3D(object):
+    def __init__(self, n, d, precision='double', verbose=False,
+                 cuda_threads=64, cuda_cwr_loops=32):
+        if d > n:
+            raise ValueError("d ({d}) must be less or equal n ({n})".format(
+                d=d, n=n))
+        self.n = n
+        self.d = d
+        self.precision = precision
+        self.verbose = verbose and rank == 0
+        self.cuda_threads = cuda_threads
+        self.cuda_cwr_loops = cuda_cwr_loops
+        self.map_iterations = self._create_iteration_table()
 
-    .. centered::  Xi ~ Uniform(0, 1),
+    @property
+    def num_iterations(self):
+        # map_iterations table is populated with element indices, not counts;
+        # therefore, we add 1
+        return self.map_iterations[:, -1].sum() + 1
+
+    def _create_iteration_table(self):
+        # do not use numpy arrays - they are limited to uint64
+        map_iterations = [list(range(self.n))]
+        for row_id in range(1, self.d):
+            prev_row = map_iterations[row_id - 1]
+            curr_row = [0] * (row_id + 1)
+            for col_id in range(row_id + 1, self.n):
+                cumsum = prev_row[col_id] + curr_row[-1]
+                curr_row.append(cumsum)
+            map_iterations.append(curr_row)
+        # here we can wrap the resulting array in numpy:
+        # if at least one item is greater than 2<<63 - 1,
+        # the data type will be set to 'object'
+        map_iterations = np.vstack(map_iterations)
+        return map_iterations
+
+    def _next_sequence_sorted(self, iteration):
+        # an alternative implementation to naive for-loop iteration when the
+        # MPI size is large. However, it's not clear under which circumstances,
+        # if any, there is a benefit. That's why this function is not used.
+        sequence_sorted = []
+        element = self.n - 1
+        for row in range(self.d - 1, -1, -1):
+            map_row = self.map_iterations[row]
+            while element > row and iteration < map_row[element]:
+                element -= 1
+            iteration -= map_row[element]
+            sequence_sorted.append(element + 1)
+        return tuple(sequence_sorted)
+
```

Comment:
I guess you're planning to remove this?

Reply:
I've included this function as one-to-one correspondence with https://github.com/INM-6/elephant/blob/cuda/asset/elephant/asset/asset.template.cu#L85 and showed in tests how one can use it. Two reasons to keep it:
1. It helps to understand the CUDA code.
2. Heat framework could utilize this method in their heat-accelerated ASSET, if their framework allows users to operate directly with CUDA kernels that require iterations to be independent by nature.
I've spent two weeks agonizing that ASSET can be rewritten to get rid of its sequential processing, which ruins CUDA acceleration. This function proves the idea is correct.